### PR TITLE
Enhancement: Ensure all contract objects return a txHash

### PR DESCRIPTION
**Pull request by CruzMolina**
This PR addresses #2291 by making sure a `transactionHash` is attached to a contract object when it's successfully returned. Ensures `MyContract.at`, `MyContract.new` & `MyContract.deployed` calls return the same expected form of contract object.

### CODE REVIEW

File: packages/contract/lib/contract.js [hunk 1]
```diff
@@ -20,13 +20,6 @@ if (typeof Web3 === "object" && Object.keys(Web3).length === 0) {
     var instance = this;
     var constructor = instance.constructor;
```
Comment (eggplantzzz) [inline packages/contract/lib/contract.js]: Will this be a breaking change removing this functionality? I'm not sure if it is useful for external use or if it is even documented.

Comment (CruzMolina) [reply packages/contract/lib/contract.js]: Good question. It's at least an internal/module-specific breaking change. I think the benefit of normalizing the documented & exposed API behavior outweighs the rare possible usage of something like `const newInstance = new SimpleStorage(SimpleStorage.address)` vs `const newInstance = SimpleStorage.at(SimpleStorage.address)`.

I don't think we really want or expect people to do the former.

Comment (gnidan) [reply packages/contract/lib/contract.js]: It is a breaking change! Also, there's https://en.m.wikipedia.org/wiki/Robustness_principle, which says to be liberal in what inputs you accept.

I am against removing this.

Comment (CruzMolina) [reply packages/contract/lib/contract.js]: Good point. Will rework the code to be non-breaking.

Comment (eggplantzzz) [reply packages/contract/lib/contract.js]: By the way, I'm not against your renaming the input to this function to more accurately describe what it is. When the argument says "contract" I expect it to be a contract, not a string representation of the address. This is the kind of stuff that makes the code super confusing and hard to read. :)

File: packages/contract/lib/contract.js [hunk 2]
```diff
@@ -20,13 +20,17 @@ if (typeof Web3 === "object" && Object.keys(Web3).length === 0) {
...
+    // in this case contract is a contract address string
     if (typeof contract === "string") {
-      var web3Instance = new constructor.web3.eth.Contract(constructor.abi);
-      web3Instance.options.address = contract;
-      contract = web3Instance;
+      const contractInstance = new constructor.web3.eth.Contract(
+        constructor.abi
+      );
+      contractInstance.options.address = contract;
+      if (constructor._json.networks[constructor.network_id])
```
Comment (haltman-at) [inline packages/contract/lib/contract.js]: May want braces here to avoid future problems

Comment (CruzMolina) [reply packages/contract/lib/contract.js]: >Or would it be possible to only populate transactionHash when the address passed to at matches the address in the artifact, in which case you can just look up the transaction hash there?

This makes sense!

Comment (CruzMolina) [reply packages/contract/lib/contract.js]: Actually, thinking about this some more. `MyContract.at('0x...')` will *only* return with a `transactionHash` if it's been deployed with Truffle to the current network (via `network_id`) that you're connected/connecting to.

So for example, if you used Truffle to migrate `MyContract` to Ropsten, but then try to get an instance of `MyContract` that you deployed to Rinkeby using something _other_ than Truffle (i.e. Remix), that contract's information won't be stored in the artifacts, so if you call `MyContract.at()` to the address on Rinkeby, the `transactionHash` won't be available.

Comment (haltman-at) [reply packages/contract/lib/contract.js]: But there could be multiple instances of a contract on the same network.  Suppose you used Truffle to deploy `MyContract` to Rinkeby, and then deployed another copy of `MyContract` to Rinkeby by a different means; say for instance that it was created as part of a transaction and you got the address from an event.  Using `MyContract.at()` on that second address shouldn't give you a contract with the `transactionHash` of the first.

Comment (CruzMolina) [reply packages/contract/lib/contract.js]: Good point!

File: packages/contract/lib/contract.js [hunk 3]
```diff
@@ -20,13 +20,20 @@ if (typeof Web3 === "object" && Object.keys(Web3).length === 0) {
...
-      contract = web3Instance;
+      const contractInstance = new constructor.web3.eth.Contract(
+        constructor.abi
+      );
+      contractInstance.options.address = contract;
+      if (
+        constructor._json.networks[constructor.network_id] &&
+        constructor._json.networks[constructor.network_id].address === contract
+      )
+        contractInstance.transactionHash = constructor.transactionHash;
```
Comment (haltman-at) [inline packages/contract/lib/contract.js]: Again, might want braces here to prevent confusion

**Pull request merged**